build/__main__.py

```diff
@@ -27,25 +29,33 @@ def _error(msg, code=1):  # type: (str, int) -> None  # pragma: no cover
     exit(code)
 
 
-def build(srcdir, outdir, distributions, skip_dependencies=False):  # type: (str, str, List[str], bool) -> None
+def build(srcdir, outdir, distributions, isolation=True, skip_dependencies=False):
+    # type: (str, str, List[str], bool, bool) -> None
     '''
     Runs the build process
 
     :param srcdir: Source directory
     :param outdir: Output directory
     :param distributions: Distributions to build (sdist and/or wheel)
+    :param isolation: Isolate the build in a separate environment
     :param skip_dependencies: Do not perform the dependency check
     '''
     try:
         builder = ProjectBuilder(srcdir)
 
-        for dist in distributions:
-            if not skip_dependencies:
-                missing = builder.check_depencencies(dist)
-                if missing:
-                    _error('Missing dependencies:' + ''.join(['\n\t' + dep for dep in missing]))
-
-            builder.build(dist, outdir)
+        if isolation:
+            with pep517.envbuild.BuildEnvironment() as env:
```

Comment:
I just had a conversation that made me realize that this may actually not be sufficient to achieve proper isolation. I was under the impression that `BuildEnvironment()` would create an *isolated* build environment, but it actually adds a layer on top of your existing environment. For example, if I create a project with the following `setup.py`:

```python
from setuptools import setup

try:
    import attr
    raise ValueError("Attrs should not be available!")
except ImportError:
    pass

setup()
```

If I run `python -m build` on the project without installing `attrs` it works fine, but if I `pip install attrs` into the same environment that `build` is installed in, it raises the exception.

That's especially troubling because `build` itself has runtime dependencies, so any undeclared dependencies on `packaging`, `pep517`, `pyparsing`, `toml`, etc, would go unnoticed.

I don't know if we should try and fix this here or in `pep517`. I assume @gaborbernat will have a better understanding of the options than I do.

Reply:
AFAIK (@takluyver would know for sure) it's the caller responsibility to ensure that on the host python only the standard library is available. So I'd say the responsibility would be on this lib to fix this. In practice, this would imply that you would need to remove from ``sys.path`` all paths that are not stdlib (you can get this via https://docs.python.org/3/library/sysconfig.html#sysconfig.get_path and passing in stdlib).

Reply:
"The caller" as in the user of `python-build`, or "The caller" as in `python-build`?

I think the way it is now we're basically doing the equivalent of building a virtualenv with `--system-site-packages`, which is not a good default, so I feel strongly that we should *not* be doing that. How does `virtualenv` achieve this?

I suppose this explains why `pip` has their own logic for build isolation and doesn't directly use `pep517`'s. We may want to copy them (though their approach does bring in [at least one problem](https://github.com/pypa/pip/issues/6264)).

Reply:
> How does `virtualenv` achieve this?

It knows the interpreter deeply to be able to tell how to achieve. There's no single global answer, various interpreters do it differently. Here's a list https://github.com/pypa/virtualenv/blob/master/src/virtualenv/discovery/py_info.py#L35-L103 of options that can influence things. Either pick pips way or pull in virtualenv package to create a virtual environment the way I see it. I agree we should not pull in global site package.

Reply:
We're thinking of removing `BuildEnvironment` from `pep517` (along with all the higher-level code that uses it) - it's not very good, as you've noticed, and it distracts people from the useful part of the pep517 package. See https://github.com/pypa/pep517/pull/83

I wrote `BuildEnvironment` and the code using it as a kind of example of the simplest possible PEP 517 frontend that puts all the pieces together (installing dependencies, calling the hooks with those dependencies available). The spec only says that the build environment SHOULD be isolated, so that 'overlay' model is technically compliant, but it's not best practice.

Reply:
I think having this high level code would be in-scope for this project, we just need to be careful with what we choose to export as public API.

So I propose for now to move `BuildEnvironment` here as private API, and then in the future think about exporting it.

> I wrote `BuildEnvironment` and the code using it as a kind of example of the simplest possible PEP 517 frontend that puts all the pieces together (installing dependencies, calling the hooks with those dependencies available).

I have expressed this opinion previously, but I'll reiterate as I think it's relevant. Nothing you consider an "example" should be ever exported publicly, being this API, CLI, etc. Once you make something public, people will use it. Providing code as an example forces people to copy it, tailor it to their needs, and most importantly, lets you off the hook :smile:.

Reply:
Yeah, I'm right with you there. I must have been feeling rather gung-ho when I decided to make it part of the library. ;-)

Reply:
I don't think the issue here was if it's private or not. The issue is that it's not working, so we should use something else, which can be private. I'd say perhaps virtualenv, but I'm happy with venv too.

Reply:
I don't think the current `pep517` code should be hard to fix up, we just need to adjust `sys.path`, no? I think `virtualenv` would be the direction forward if we needed to deal with custom interpreters, but we don't. I would like to avoid adding a new dependency.

Reply:
> I don't think the current `pep517` code should be hard to fix up, we just need to adjust `sys.path`, no?

Well sure, in theory, you *just** need to fix the sys.path; however, in practice, it's non-trivial how to fix it up. Things differ per python implementation what needs to stay and what needs to go. virtualenv contains this complex logic. Hence why suggesting we use the slower venv on python 3, as that does not pull in any dependency 🤷